src/Symfony/Component/Form/FormRegistry.php

```diff
@@ -97,13 +100,30 @@ public function getType($name)
      */
     private function resolveAndAddType(FormTypeInterface $type)
     {
+        if (isset($this->checkedTypes[$type->getName()])) {
+            $types = implode(' > ', array_merge(array_keys($this->checkedTypes), array($type->getName())));
+            throw new LogicException(sprintf('Circular reference detected for form "%s" (%s).', $type->getName(), $types));
+        }
+
+        $this->checkedTypes[$type->getName()] = true;
+
         $parentType = $type->getParent();
 
         if ($parentType instanceof FormTypeInterface) {
+            if ($parentType->getName() === $type->getName()) {
+                $this->checkedTypes = array();
```

Comment:
I think we should still use a try/catch emulating finally, so that the code can recover from any unexpected exception